hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationContext.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl.operationservice.impl;
+
+import com.hazelcast.instance.Node;
+import com.hazelcast.internal.cluster.ClusterClock;
+import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.internal.partition.InternalPartitionService;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.internal.util.counters.MwCounter;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.ConnectionManager;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.impl.executionservice.InternalExecutionService;
+import com.hazelcast.spi.impl.operationexecutor.OperationExecutor;
+import com.hazelcast.util.executor.ManagedExecutorService;
+
+/**
+ * The InvocationContext contains all 'static' dependencies for an Invocation; dependencies that don't change between
+ * invocations. All invocation specific dependencies/settings are passed in through the constructor of the invocation.
+ *
+ * This object should have no functionality apart from providing dependencies. So no methods should be added to this class.
+ *
+ * The goals of the InvocationContext are:
+ * <ol>
+ * <li>prevent a.b.c.d call in the invocation by pulling all dependencies in the InvocationContext</li>
+ * <li>reduce the need on having a cluster running when testing Invocations.</li>
+ * <lI>removed dependence on NodeEngineImpl. Only NodeEnigne is needed to set on Operation</lI>
+ * <li>reduce coupling to Node. Most calls point now direct to the right dependency, instead of needing
+ * to go through node/NodeEngine. This makes it easier in the future to get rid of Node</li>
+ * </ol>
+ */
+class InvocationContext {
```

Comment:
I'm very unhappy with the style of this class :(
* CheckStyle and SonarQube will create visibility errors for each field
* you'll never know if the context has been constructed completely
* every user of the context can overwrite any field, even by accident!
* everyone can overwrite any field at any time
* did I mention that everyone can overwrite... we should not have this in the SPI!!!
* refactoring will be very painful with this code style (been there, done that with Simulator)
* refactoring is risky on that code style (e.g. you add a new field and no compiler error if you forget to initialize a field in a test, you don't know about)
* this can lead to NPE in production code or tests
* this can lead to silent unexpected behavior if there are null checks and "failover" code paths are used (been there, done that with Simulator)
* the 20 something line construction looks not better than a constructor with all fields
* a constructor with so many parameters will cause a CheckStyle issue as well, but it's a single one for a method instead of one for each field (and the suppression is totally fine for me, since it's a context class)
* IntelliJ will write all the boilerplate code for you, even if you have to change anything
* IntelliJ has shortcuts to change the order of fields, constructor parameters etc.
* InjtelliJ will most probably do everything for you which may lead you to this style
* ...

So `final` for mandatory fields is a must for me. A constructor is a must for me. `package private` is a must for me, `private` and getter are a nice to have for me (although I cannot understand why getters are bad, IntelliJ to the rescue).